lichess-bot.py

```diff
@@ -32,6 +32,8 @@
 
 terminated = False
 
+out_of_online_opening_book_moves = 0
```

Comment:
This needs to be a dictionary from game IDs to failed lookup counts. As it is now, the first game to fail to find enough moves will prevent all subsequent games from using the opening book. Simultaneous games will exhaust the opening book faster.

To keep the dictionary from growing arbitrarily large as games accumulate, you can watch for the `gameFinish` state in the `start()` function to remove entries from the dictionary.

Reply:
Actually, a `defaultdict` with `default_factory=int` is even better to avoid the boilerplate `if game.id not in out_of_online_opening_book_moves:`.

Reply:
I moved it inside `play_game` to also avoid `global`.

Reply:
The problem here is that the counts are lost if `play_game()` exits in the middle of a game. This could happen in a correspondence game or because of connection issues.

If you want to avoid globals, put the dictionary version inside `start()` before the main loop and pass it into `play_game()`. Since dictionaries are mutable, you can modify them and the modifications will persist after the function call, so there's no need to return it.

Also, [Counter](https://docs.python.org/3/library/collections.html#collections.Counter) is probably the simplest data structure to use for this.